stats: index out of range crash when forecasting

(I was mistakenly running using the binary from late June, about on fa47de054aa94a7207f3140cc8dfd87b36f8ec03.)

I hit this crash on one node when running `unoptimized-query-oracle/disable-rules=half/seed-multi-region` with #108972
```
panic: runtime error: index out of range [0] with length 0

goroutine 91801 [running]:
panic({0x5dcdee0, 0xc00ec84d98})
        GOROOT/src/runtime/panic.go:987 +0x3ba fp=0xc00ab1f280 sp=0xc00ab1f1c0 pc=0x49beda
runtime.goPanicIndex(0x0, 0x0)
        GOROOT/src/runtime/panic.go:113 +0x7f fp=0xc00ab1f2c0 sp=0xc00ab1f280 pc=0x499cff
github.com/cockroachdb/cockroach/pkg/sql/stats.quantile.fixMalformed({0xc007cfb000?, 0xb0, 0x160})
        github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/quantile.go:806 +0xd3b fp=0xc00ab1f468 sp=0xc00ab1f2c0 pc=0x22d589b
github.com/cockroachdb/cockroach/pkg/sql/stats.predictHistogram({0x7812f40, 0xc0052b48a0}, {0xc00315fa60, 0x3, 0x3?}, 0xc00ec84d20?, 0x3fee666666666666, 0x4053c00000000000)
        github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/forecast.go:413 +0x67c fp=0xc00ab1f608 sp=0xc00ab1f468 pc=0x22cbf7c
github.com/cockroachdb/cockroach/pkg/sql/stats.forecastColumnStatistics({0x7812f40, 0xc0052b48a0}, 0xc008780970?, {0xc00315fa60, 0x3, 0x4}, {0xc00bc29c00?, 0xc00bc29cc8?, 0x0?}, 0x3fee666666666666)
        github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/forecast.go:275 +0xca5 fp=0xc00ab1f968 sp=0xc00ab1f608 pc=0x22ca8a5
github.com/cockroachdb/cockroach/pkg/sql/stats.ForecastTableStatistics({0x7812f40, 0xc0052b48a0}, 0xc00c123280?, {0xc00a750780?, 0x9, 0xa})
        github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/forecast.go:115 +0x52d fp=0xc00ab1fc60 sp=0xc00ab1f968 pc=0x22c99cd
github.com/cockroachdb/cockroach/pkg/sql/stats.(*TableStatisticsCache).getTableStatsFromDB(0xc002127080, {0x7812f40, 0xc0052b48a0}, 0x7812f40?, 0x1)
        github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/stats_cache.go:826 +0x60a fp=0xc00ab1fe20 sp=0xc00ab1fc60 pc=0x22dcc0a
github.com/cockroachdb/cockroach/pkg/sql/stats.(*TableStatisticsCache).refreshCacheEntry.func1(0xc002127080, {0x7812f40, 0xc0052b48a0}, 0x4b6dd00?, 0xc0?, 0xc00ab1ff50, 0xc00ab1ff30)
        github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/stats_cache.go:451 +0x14e fp=0xc00ab1fec0 sp=0xc00ab1fe20 pc=0x22da56e
github.com/cockroachdb/cockroach/pkg/sql/stats.(*TableStatisticsCache).refreshCacheEntry(0xc002127080, {0x7812f40, 0xc0052b48a0}, 0x1b44500?, {0xc009583000?, 0xfc7c50?, 0xc0?})
        github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/stats_cache.go:453 +0x219 fp=0xc00ab1ff80 sp=0xc00ab1fec0 pc=0x22da239
github.com/cockroachdb/cockroach/pkg/sql/stats.(*TableStatisticsCache).refreshTableStats.func1()
        github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/stats_cache.go:480 +0x74 fp=0xc00ab1ffe0 sp=0xc00ab1ff80 pc=0x22daa14
runtime.goexit()
        GOROOT/src/runtime/asm_amd64.s:1594 +0x1 fp=0xc00ab1ffe8 sp=0xc00ab1ffe0 pc=0x4d1061
created by github.com/cockroachdb/cockroach/pkg/sql/stats.(*TableStatisticsCache).refreshTableStats
        github.com/cockroachdb/cockroach/pkg/sql/stats/pkg/sql/stats/stats_cache.go:478 +0x1a5
```
Here I'm attaching somewhat reduced version of the artifacts (so that it would fit under 25MiB limit of GitHub). I didn't try to reproduce this.

[artifacts.zip](https://github.com/cockroachdb/cockroach/files/12384607/artifacts.zip)


Jira issue: CRDB-30757

--- a/pkg/sql/stats/BUILD.bazel
+++ b/pkg/sql/stats/BUILD.bazel
@@ -52,6 +52,7 @@ go_library(
         "//pkg/util/hlc",
         "//pkg/util/intsets",
         "//pkg/util/log",
+        "//pkg/util/log/logcrash",
         "//pkg/util/mon",
         "//pkg/util/protoutil",
         "//pkg/util/stop",

--- a/pkg/sql/stats/forecast.go
+++ b/pkg/sql/stats/forecast.go
@@ -27,6 +27,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/types"
 	"github.com/cockroachdb/cockroach/pkg/util/errorutil"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/log/logcrash"
 	"github.com/cockroachdb/errors"
 	"github.com/cockroachdb/redact"
 )
@@ -272,7 +273,7 @@ func forecastColumnStatistics(
 	// histogram. NOTE: If any of the observed histograms were for inverted
 	// indexes this will produce an incorrect histogram.
 	if observed[0].HistogramData != nil && observed[0].HistogramData.ColumnType != nil {
-		hist, err := predictHistogram(ctx, observed, forecastAt, minRequiredFit, nonNullRowCount)
+		hist, err := predictHistogram(ctx, sv, observed, forecastAt, minRequiredFit, nonNullRowCount)
 		if err != nil {
 			// If we did not successfully predict a histogram then copy the latest
 			// histogram so we can adjust it.
@@ -358,6 +359,7 @@ func forecastColumnStatistics(
 // predictHistogram tries to predict the histogram at forecast time.
 func predictHistogram(
 	ctx context.Context,
+	sv *settings.Values,
 	observed []*TableStatistic,
 	forecastAt float64,
 	minRequiredFit float64,
@@ -410,7 +412,12 @@ func predictHistogram(
 	// Construct a linear regression model of quantile functions over time, and
 	// use it to predict a quantile function at the given time.
 	yₙ, r2 := quantileSimpleLinearRegression(createdAts, quantiles, forecastAt)
-	yₙ = yₙ.fixMalformed()
+	var err error
+	yₙ, err = yₙ.fixMalformed()
+	if err != nil {
+		logcrash.ReportOrPanic(ctx, sv, "unable to fix malformed stats quantile: %v", yₙ)
+		return histogram{}, err
+	}
 	log.VEventf(
 		ctx, 3, "forecast for table %v columns %v predicted quantile %v R² %v",
 		tableID, columnIDs, yₙ, r2,

--- a/pkg/sql/stats/quantile.go
+++ b/pkg/sql/stats/quantile.go
@@ -105,7 +105,7 @@ var zeroQuantile = quantile{{p: 0, v: 0}, {p: 1, v: 0}}
 // If you are introducing a new histogram version, please check whether
 // makeQuantile and quantile.toHistogram need to change, and then increase the
 // hard-coded number here.
-const _ uint = 2 - uint(histVersion)
+const _ = 2 - uint(histVersion)
 
 // canMakeQuantile returns true if a quantile function can be created for a
 // histogram of the given type. Note that by not supporting BYTES we rule out
@@ -656,10 +656,10 @@ func (q quantile) integrateSquared() float64 {
 // This function fixes the negative slope pieces by "moving" p from the
 // overlapping places to where it should be. No p is lost in the making of these
 // calculations.
-func (q quantile) fixMalformed() quantile {
+func (q quantile) fixMalformed() (quantile, error) {
 	// Check for the happy case where q is already well-formed.
 	if q.isWellFormed() {
-		return q
+		return q, nil
 	}
 
 	// To fix a malformed quantile function, we recalculate p for each distinct
@@ -803,6 +803,9 @@ func (q quantile) fixMalformed() quantile {
 		// "below" our horizontal line (<= v) and ends "above" our horizontal line
 		// (> v) and we always add two intersectionPs for "double roots" touching
 		// our line from above (one endpoint > v, one endpoint = v).
+		if len(intersectionPs) == 0 {
+			return q, errors.New("unable to fix malformed stats quantile")
+		}
 		lessEqP := intersectionPs[0]
 		for j := 1; j < len(intersectionPs); j += 2 {
 			lessEqP += intersectionPs[j+1] - intersectionPs[j]
@@ -818,7 +821,7 @@ func (q quantile) fixMalformed() quantile {
 			fixed = append(fixed, quantilePoint{p: lessEqP, v: val})
 		}
 	}
-	return fixed
+	return fixed, nil
 }
 
 // isWellFormed returns true if q is well-formed (i.e. is non-decreasing in v).

--- a/pkg/sql/stats/quantile_test.go
+++ b/pkg/sql/stats/quantile_test.go
@@ -1348,7 +1348,10 @@ func TestQuantileOps(t *testing.T) {
 			if intSqMult0 != 0 {
 				t.Errorf("test case %d incorrect intSqMult0 %v expected 0", i, intSqMult0)
 			}
-			fixed := tc.q.fixMalformed()
+			fixed, err := tc.q.fixMalformed()
+			if err != nil {
+				t.Errorf("test case %d has an unexpected error: %s", i, err)
+			}
 			if !reflect.DeepEqual(fixed, tc.fixed) {
 				t.Errorf("test case %d incorrect fixed %v expected %v", i, fixed, tc.fixed)
 			}
@@ -1407,7 +1410,10 @@ func TestQuantileOpsRandom(t *testing.T) {
 			if intSqMult0 != 0 {
 				t.Errorf("seed %v quantile %v incorrect intSqMult0 %v expected 0", seed, q, intSqMult0)
 			}
-			fixed := q.fixMalformed()
+			fixed, err := q.fixMalformed()
+			if err != nil {
+				t.Errorf("seed %v quantile %v had an unexpected error: %s", seed, q, err)
+			}
 			intSqFixed := fixed.integrateSquared()
 			// This seems like it should run into floating point errors, but it hasn't
 			// yet, so yay?

--- a/pkg/sql/stats/simple_linear_regression_test.go
+++ b/pkg/sql/stats/simple_linear_regression_test.go
@@ -286,7 +286,11 @@ func TestQuantileSimpleLinearRegression(t *testing.T) {
 	for i, tc := range testCases {
 		t.Run(strconv.Itoa(i), func(t *testing.T) {
 			yn, r2 := quantileSimpleLinearRegression(tc.x, tc.y, tc.xn)
-			yn = yn.fixMalformed()
+			var err error
+			yn, err = yn.fixMalformed()
+			if err != nil {
+				t.Errorf("test case %d had an unexpected error: %s", i, err)
+			}
 			if !reflect.DeepEqual(yn, tc.yn) {
 				t.Errorf("test case %d incorrect yn %v expected %v", i, yn, tc.yn)
 			}

